src/fs/metadata_log_operations/write.hh

```diff
@@ -296,13 +300,21 @@ private:
             case metadata_log::append_result::TOO_BIG:
                 return make_exception_future<size_t>(cluster_size_too_small_to_perform_operation_exception());
             case metadata_log::append_result::NO_SPACE:
-                _metadata_log._cluster_allocator.free(cluster_id);
                 return make_exception_future<size_t>(no_more_space_exception());
             case metadata_log::append_result::APPENDED:
                 _metadata_log.memory_only_disk_write(_inode, file_offset, cluster_disk_offset, write_len);
                 return make_ready_future<size_t>(write_len);
             }
             __builtin_unreachable();
+        }).then([this, cluster_id](size_t write_len) -> size_t {
+            if (write_len != _metadata_log._cluster_size) {
+                _metadata_log._cluster_allocator.free(cluster_id);
+                return 0;
+            }
+            return write_len;
+        }).handle_exception([this, cluster_id](std::exception_ptr ptr) {
+            _metadata_log._cluster_allocator.free(cluster_id);
+            return make_exception_future<size_t>(std::move(ptr));
```

Comment:
Should this be part of adding read-only mode or not?

Reply:
It is added because handling another exception would result in copy-pasting code that returns cluster_id to cluster_allocator. I will fix the bug with freeing cluster_id in commit introducing write operation but I would leave that refactor here because it was added because of read-only mode.

Reply:
So it fixes a bug and is orthogonal with the addition of read only mode... Thus I would like to see it as a separate FIXUP commit.

Reply:
It fixes a bug as a side effect of refactor caused by an additional exception thrown by adding read-only mode.

Reply:
And you will see additional fixup commit but without that refactor.